### DISCUSSION TITLE
Apple iOS Secure-Transport deprecated, is there plan to port to use Network.framework

### DISCUSSION THREAD

**Discussion by sbelbin** [curl/curl#15759] [General]
Apple has deprecated **Secure-Transport** capabilities since iOS 13.0 (which is several years by now) with a recommendation of porting to. use **Network.framework** instead.

Therefore building libCURL with the **--with-secure-transport** doesn't work out-of-the-box when using Apple's iOS SDK beyond 13.0. I did it build but required a series of kludges in order to resolve linker errors of the missing API from Secure-Transport.

I'm developing an iOS app that requires access to CA certificates stored in the operating system's Keychain, so that administrators of these iOS devices are responsible to maintain the devices' CA certificates in Configuration Profiles under Settings | VPN & Device Management. In addition to updates of those that are included whenever updating the iOS that are provided by Apple.

Additionally, this app has a requirement of using technology that isn't deprecated (or going to be deprecated in the foreseeable future) and this includes its dependencies such as libCURL built with iOS Secure-Transport **isn't** a viable choice for us.

Thus, I'd like to know if anyone is aware of means to build libCURL with options so that it can access an iOS device's CA certificates are stored in the operating system's Keychain using a solution that isn't deprecated. Or if someone is implementing on a solution for iOS (such as using Network.framework)?

Regards,
Steven Belbin

**Comment by danielgustafsson** [comment]
I don't know of any such development, but making is possible to use certificates from the Keychain in a non-SecureTransport TLS lib doesn't seem entirely impossible (the devil being in the detail).

If I understand things correctly the main issue is that the new Network.framework API and objectmodel isn't a good fit for libcurl and would thus require a substantial effort to support.

**Comment by bagder** [comment]
- There is no known effort to support **Network.framework**, and it seems to be a quirky interface to work with so it seems unlikely to happen.
- We are going to remove Secure Transport support in May 2025, primarily because its lack of support for TLS 1.3.
- Adding support for accessing and using an iOS device's CA certificates sounds like a worthy task. Someone should do it.

**Reply by sbelbin** [reply]
Hi @bagder 

I appreciate the quick response (also to @danielgustafsson) and pointing that removal of Secure Transport in libCURL is slated for the next upcoming May in 2025.

I'm of the opinion that as part of the removal Secure Transport needs to include the  planning of its replacement added to CURL (libCURL) project's roadmap.

Given that the number of devices the number of iOS (& its variants such as iWatchOS, iVisionOS) are projected to increase in the upcoming years, then work for replacement for Secure-Transport will be appreciated by the number apps running on those devices that rely on the CA certificates within Keychain.

Otherwise, those apps will break as they port to using later revisions of libCURL (after May 2025) in order benefit using a new capability, performance improvements or requiring a fix of critical issues (i.e. CVE) that are available in those later revisions.


Unfortunately, at this time I can't take on a task of trying to implement a replacement of Secure-Transport since I've got time constraints in terms of professional and personal lives that require my attention.

However, I'm willing to set aside some time doing the following:
- Participate in discussions.
- Review/guide the plan to replace as well as its implementation.
- Being an alpha tester.
- Help to diagnosis and to troubleshoot issues.

Regards,
Steven

**Reply by danielgustafsson** [reply]
> I'm of the opinion that as part of the removal Secure Transport needs to include the planning of its replacement added to CURL (libCURL) project's roadmap.

The curl project is not responsible for deprecation choices made by multibilliondollar companies, nor are required to react to them in other ways than match the deprecation by removal of support. I agree that supporting certificates in Keychain would be cool in non-SecureTransport libcurl, but it's far from trivial. The point of vaults is to not remove artifacts from them, in order to allow for example OpenSSL to use a cert/key you'd have to copy them to disk etc (according to my understanding, I might be wrong). I'm sure there are ways forward, but it (probably) won't be a trivial patch.
 
> Given that the number of devices the number of iOS (& its variants such as iWatchOS, iVisionOS) are projected to increase in the upcoming years, then work for replacement for Secure-Transport will be appreciated by the number apps running on those devices that rely on the CA certificates within Keychain.
> 
> Otherwise, those apps will break as they port to using later revisions of libCURL (after May 2025) in order benefit using a new capability, performance improvements or requiring a fix of critical issues (i.e. CVE) that are available in those later revisions.

Great, sounds like there's a lot of stakeholders here who can help fund the work (developer time or money), else maybe it wasn't such a priority after all? An open source supply chain is one which producers (in this case, app developers) can directly influence and empower as opposed to many other industries - it's incredibly powerful but it's a shared responsibility.
 
> Unfortunately, at this time I can't take on a task of trying to implement a replacement of Secure-Transport since I've got time constraints in terms of professional and personal lives that require my attention.

That's a pretty common constraint, and when coupled with not having an app in any Apple appstore becomes an even higher bar to cross.

Ideally something can happen, but nothing here is new, this has been known for a long time and no one has seemed terribly keen to fix it thus far.

**Reply by bagder** [reply]
> the number of devices the number of iOS are projected to increase in the upcoming years

I just want to concur with @danielgustafsson that this projected future actually does not matter to us.

We would **love** to support all those latest things and features on those devices and operating systems, but no users or developers on those systems should just assume that someone else will do the work (for free).

Make it happen if you need it to happen.

If not, expect to see Secure Transport get dropped and since no one has even tried to do anything about it yet, I presume that also means that come May 2025, libcurl users on apple devices will get a brutal awakening and realization they won't like at all.

Apple themselves have a patch applied on their libressl build to make it use the system CA certs somehow, so it is certainly possible. We know they do, because they told me: https://daniel.haxx.se/blog/2024/03/08/the-apple-curl-security-incident-12604/. That's what the curl tool on apple OSes use.

**Reply by nickzman** [reply]
FWIW, I once started on porting libcurl to the Network framework. After realizing that implementation and maintenance was not going to be trivial (since we'd now have to support two completely different socket implementations), I abandoned my efforts.

Last time I used it, ironically, the Network framework used Secure Transport for TLS. I don't know if it has since been moved to BoringSSL or something else.

I would prefer it if Apple would continue to maintain Secure Transport, but I would understand if it's time for it to go. And I wrote the Secure Transport interface in libcurl.

**Reply by sbelbin** [reply]
I'd appreciate you sharing your experiences of attempting to integrate the **Network.framework** so that it can be used in **libcurl**.

>I also find it humorous about the **Network.framework** uses(ed) **Secure-Transport**. Given that key-point was weening developers from using **Secure-Transport** due to various concerns that Apple have about it.

I looked at the **Network.framework** and I understood the amount of effort required to build a shim (possibly using as a C ABI) allowing **libcurl** to interoperate with it.

> Not to mention the added complexity of supporting another socket paradigm. Something of which I was unaware about, since I'm familiar with **libcurl**'s external API while having only a basic understanding about it's internals.

As I've mentioned in another comment (see below), for the time being I'll investigate the possibly using **libreSSL** as potential option for the needs of the iOS app that I'm implementing.

**Comment by bagder** [comment]
Cory Benfield mention this on Mastodon:

For validating certs using the platform trust store while not using the platform TLS library, we have an example you can use: https://github.com/apple/swift-nio-ssl/blob/main/Sources/NIOSSL/SecurityFrameworkCertificateVerification.swift.

Swift code, but I’m sure a dedicated contributor could translate it to C.

**Comment by sbelbin** [comment]
Thanks to all for sharing your thoughts & including insights, knowledge & experiences about the topic of Secure-Transport and using certificates from the Keychain.

The article that @bagder article posted about **libreSSL** behaviour on the macOS platform to apply a fallback of fetching the certificates from the Keychain just might be good enough for the needs of the app that I'm implementing. I'll retry building **libreSSL** for iOS platform from source code to determine if it exhibits the same behaviours as for the macOS platform.

> I'd previously attempted to build it for iOS platform and it didn't work. However, it's possible that I didn't provide a particular options or settings to activate this behaviour that platform.

Should I find the options/settings where the iOS app behaves the same as that on a macOS platform, then I'll post those details here so that others can opt applying this approach into their iOS apps.

> For your information. The iOS app that I'm implementing transmits video + audio acquired from an iOS device's camera + microphone to transmit them to an AWS KVS video stream. The app uses Amazon's AWS KVS GStreamer **kvssink** plugin thatdepends on **libCURL** for the exchanging with the AWS KVS endpoints.

Regards,
Steven Belbin

**Comment by conradev** [comment]
Hi everyone! I started looking into this because I wanted to use curl on watchOS. Network.framework is actually required on watchOS because the POSIX socket APIs are unavailable. I actually went and got something working pretty well:

https://github.com/conradev/curl

The APIs are quite compatible, actually, and I've been meaning to open a draft PR. It has nice features like TLS 1.3 and ECH. I still need to implement control over cipher suite selection and a handful of other things.

Network.framework has two APIs: an object-oriented Objective-C API and a C API. I used the C API and it is a nice fit for curl, but it does use blocks in the API. I also share code with the existing SecureTransport backend for things like collecting certificates.

edit: I have a lot of rebasing to do, so I can open a PR after I rebase

**Comment by icing** [comment]
Hi Conrad,

> Am 02.03.2025 um 08:33 schrieb Conrad Kramer ***@***.***>:
> 
> 
> Hi everyone! I started looking into this because I wanted to use curl on watchOS. Network.framework is actually required on watchOS because the POSIX socket APIs are unavailable. I actually went and got something working pretty well:
> https://github.com/conradev/curl
> The APIs are quite compatible, actually, and I've been meaning to open a draft PR. It has nice features like TLS 1.3 and ECH. I still need to implement control over cipher suite selection and a handful of other things.
> Network.framework has two APIs: an object-oriented Objective-C API and a C API. I used the C API and it is a nice fit for curl, but it does use blocks in the API. I also share code with the existing SecureTransport backend for things like collecting certificates.

Had a look at your code. First, I have no idea on its usefulness inside a WatchOS application.

As a supported TLS backend in curl in general, it will not do. Your code is good and fairly readable. The problem is the network framework itself that exposes no sockets to monitor events on. Thus, multiple transfers in parallel will block each other at the servers' mercy. As long as there is only a single transfer at a time, it will do. If someone uses the event based scheduling, libcurl with socket callback, it will not work at all.

Also, due to the lack of sockets, IP and HTTP eyeballing will not work. Again, there is nothing lacking in your code. It is the Apple framework that does not give its user the needed capabilities. Making it a bad fit.

There are use cases, as you point out, where this is good enough. But I fail to see it as an alternate, general TLS backend in libcurl.

Cheers,
Stefan

**Reply by bagder** [reply]
Agreed. libcurl needs to setup (and own) the connection and *that* connection is the one that needs to be used by the TLS backend - in a non-blocking manner, otherwise there will be mayhem and chaos (since then we can't even know this backend speaks to the same machine).

**Reply by conradev** [reply]
Thanks for taking a look! I haven't poked around the event loop portion of libcurl much, so I really can't speak to the issues you mentioned. My loose understanding is that libcurl has a socket abstraction of some sort and a nonblocking event loop.

In the code linked above, the writes and reads are synchronous because of two reasons:
- I saw connect and connect_nonblocking (in a previous version of `master`), and that signaled to me that curl supported synchronous (blocking) backends (even if not ideal, I imagined a world where curl could make me threads, for now)
- Integrating the runloop seemed complex

That _said_, Apple has their own async runtime (GCD, the `dispatch_queue` stuff), and it should be philosophically compatible with the event loop even if the architecture is a bit different (i.e. it's not poll-based). I was going to give that a look when I had a chance. There exists some abstraction that would work here, the question to me is simply how hard it would be to build or adapt.

I might run libcurl through a language model and ask it some questions, but also I might get my bouncer working again and hop on IRC to ask more questions

Network.framework is _the_ system library for TLS on Apple's platforms. The only alternative is shipping a static TLS implementation on all of Apple's platforms for the foreseeable future which is kind of a bummer. I want curl to support the latest TLS APIs on all major platforms, Apple's included!

**Reply by conradev** [reply]
It is more akin to supporting Winsock than it is adding a new TLS backend. Dealing with the pesky closed source kernels and their networking stacks

**Comment by tmolitor-stud-tu** [comment]
I'm not sure if this is of some help since libcurl is C based, but I wrote a NSStream compatible wrapper around Apple's network.framework. Maybe that helps.

https://github.com/monal-im/Monal/blob/develop/Monal/Classes/MLStream.h
https://github.com/monal-im/Monal/blob/develop/Monal/Classes/MLStream.m

**Reply by bagder** [reply]
Sorry, but I don't see how that helps.

The network framework is at a different API level and it doesn't offer the functionality we require.

**Comment by icing** [comment]
With the support of Apple's `SecTrust` framework in `libcurl` since 8.17.0, I believe we can close this discussion. iOS/watchOS and other Apple operating system should now work well with OpenSSL/GnuTLS and SecTrust enabled in `libcurl`.

**Reply by arianvp** [reply]
Does SecTrust support client certificates?
